File: Sources/ClipboardCore/HTTPCodec.swift
import Foundation

public struct HTTPRequest: Sendable {
    public let method: String
    public let path: String
    public let headers: [String: String]
    public let body: Data
    
    public init(method: String, path: String, headers: [String: String], body: Data) {
        self.method = method
        self.path = path
        self.headers = headers
        self.body = body
    }
}

public struct HTTPResponse: Sendable {
    public let statusCode: Int
    public let headers: [String: String]
    public let body: Data
    
    public init(statusCode: Int, headers: [String: String], body: Data) {
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
    }
}

public enum HTTPCodecError: Error {
    case incomplete
    case invalidFormat
    case payloadTooLarge
}

public enum HTTPCodec {
    public static let maxBodySize = 20 * 1024 * 1024 // 20 MB

    public static func parseRequest(_ data: Data) throws -> HTTPRequest {
        guard let range = data.range(of: Data("\r\n\r\n".utf8)) else {
            throw HTTPCodecError.incomplete
        }
        
        let headData = data[..<range.lowerBound]
        let bodyData = data[range.upperBound...]
        
        guard let headString = String(data: headData, encoding: .utf8) else {
            throw HTTPCodecError.invalidFormat
        }

        var lines = headString.components(separatedBy: "\r\n")
        guard !lines.isEmpty else { throw HTTPCodecError.invalidFormat }

        let requestLine = lines.removeFirst().components(separatedBy: " ")
        guard requestLine.count == 3 else { throw HTTPCodecError.invalidFormat }

        let method = requestLine[0]
        let path = requestLine[1]

        var headers: [String: String] = [:]
        for line in lines {
            let headerParts = line.components(separatedBy: ": ")
            if headerParts.count >= 2 {
                headers[headerParts[0].lowercased()] = headerParts.dropFirst().joined(separator: ": ")
            }
        }

        if let contentLengthStr = headers["content-length"], let contentLength = Int(contentLengthStr) {
            if contentLength > maxBodySize {
                throw HTTPCodecError.payloadTooLarge
            }
            if bodyData.count < contentLength {
                throw HTTPCodecError.incomplete
            }
            return HTTPRequest(method: method, path: path, headers: headers, body: bodyData.prefix(contentLength))
        }

        return HTTPRequest(method: method, path: path, headers: headers, body: bodyData)
    }

    public static func encodeResponse(_ response: HTTPResponse) -> Data {
        let reason: String
        switch response.statusCode {
        case 200: reason = "OK"
        case 400: reason = "Bad Request"
        case 401: reason = "Unauthorized"
        case 413: reason = "Payload Too Large"
        default: reason = "Status"
        }
        
        var head = "HTTP/1.1 \(response.statusCode) \(reason)\r\n"
        var finalHeaders = response.headers
        finalHeaders["Content-Length"] = "\(response.body.count)"
        finalHeaders["Connection"] = "close"
        
        for (key, value) in finalHeaders {
            head += "\(key): \(value)\r\n"
        }
        head += "\r\n"

        var data = Data(head.utf8)
        data.append(response.body)
        return data
    }
    
    public static func parseResponse(_ data: Data) throws -> HTTPResponse {
        guard let range = data.range(of: Data("\r\n\r\n".utf8)) else {
            throw HTTPCodecError.incomplete
        }
        
        let headData = data[..<range.lowerBound]
        let bodyData = data[range.upperBound...]
        
        guard let headString = String(data: headData, encoding: .utf8) else {
            throw HTTPCodecError.invalidFormat
        }

        var lines = headString.components(separatedBy: "\r\n")
        guard !lines.isEmpty else { throw HTTPCodecError.invalidFormat }

        let statusLine = lines.removeFirst().components(separatedBy: " ")
        guard statusLine.count >= 2, let statusCode = Int(statusLine[1]) else {
            throw HTTPCodecError.invalidFormat
        }

        var headers: [String: String] = [:]
        for line in lines {
            let headerParts = line.components(separatedBy: ": ")
            if headerParts.count >= 2 {
                headers[headerParts[0].lowercased()] = headerParts.dropFirst().joined(separator: ": ")
            }
        }

        if let contentLengthStr = headers["content-length"], let contentLength = Int(contentLengthStr) {
            if contentLength > maxBodySize {
                throw HTTPCodecError.payloadTooLarge
            }
            if bodyData.count < contentLength {
                throw HTTPCodecError.incomplete
            }
            return HTTPResponse(statusCode: statusCode, headers: headers, body: bodyData.prefix(contentLength))
        }

        return HTTPResponse(statusCode: statusCode, headers: headers, body: bodyData)
    }

    public static func encodeRequest(_ request: HTTPRequest, host: String) -> Data {
        var head = "\(request.method) \(request.path) HTTP/1.1\r\n"
        var finalHeaders = request.headers
        finalHeaders["Content-Length"] = "\(request.body.count)"
        finalHeaders["Host"] = host
        finalHeaders["Connection"] = "close"
        
        for (key, value) in finalHeaders {
            head += "\(key): \(value)\r\n"
        }
        head += "\r\n"

        var data = Data(head.utf8)
        data.append(request.body)
        return data
    }
}

File: ClipboardSS.Windows/src/ClipboardSS.Core/Protocol/HttpCodec.cs
using System.Text;

namespace ClipboardSS.Core.Protocol;

public sealed record HttpRequest(
    string Method,
    string Path,
    IReadOnlyDictionary<string, string> Headers,
    byte[] Body);

public sealed record HttpResponse(
    int StatusCode,
    IReadOnlyDictionary<string, string> Headers,
    byte[] Body);

public enum HttpCodecError
{
    Incomplete,
    InvalidFormat,
    PayloadTooLarge,
}

public sealed class HttpCodecException(HttpCodecError error)
    : Exception($"HTTP codec error: {error}")
{
    public HttpCodecError Error { get; } = error;
}

public static class HttpCodec
{
    public const int MaxBodySize = 20 * 1024 * 1024;
    private static readonly byte[] HeaderTerminator = "\r\n\r\n"u8.ToArray();

    public static HttpRequest ParseRequest(ReadOnlySpan<byte> data)
    {
        var (head, body) = Split(data);
        var lines = ReadHeadLines(head);
        var requestLine = lines[0].Split(' ', StringSplitOptions.None);
        if (requestLine.Length != 3)
        {
            throw new HttpCodecException(HttpCodecError.InvalidFormat);
        }

        var headers = ParseHeaders(lines.Skip(1));
        return new HttpRequest(requestLine[0], requestLine[1], headers, ReadBody(body, headers));
    }

    public static HttpResponse ParseResponse(ReadOnlySpan<byte> data)
    {
        var (head, body) = Split(data);
        var lines = ReadHeadLines(head);
        var statusLine = lines[0].Split(' ', StringSplitOptions.None);
        if (statusLine.Length < 2 || !int.TryParse(statusLine[1], out var statusCode))
        {
            throw new HttpCodecException(HttpCodecError.InvalidFormat);
        }

        var headers = ParseHeaders(lines.Skip(1));
        return new HttpResponse(statusCode, headers, ReadBody(body, headers));
    }

    public static byte[] EncodeRequest(HttpRequest request, string host)
    {
        var headers = new Dictionary<string, string>(request.Headers, StringComparer.OrdinalIgnoreCase)
        {
            ["Content-Length"] = request.Body.Length.ToString(),
            ["Host"] = host,
            ["Connection"] = "close",
        };
        return EncodeHeadAndBody($"{request.Method} {request.Path} HTTP/1.1", headers, request.Body);
    }

    public static byte[] EncodeResponse(HttpResponse response)
    {
        var headers = new Dictionary<string, string>(response.Headers, StringComparer.OrdinalIgnoreCase)
        {
            ["Content-Length"] = response.Body.Length.ToString(),
            ["Connection"] = "close",
        };
        return EncodeHeadAndBody(
            $"HTTP/1.1 {response.StatusCode} {ReasonPhrase(response.StatusCode)}",
            headers,
            response.Body);
    }

    private static (byte[] Head, byte[] Body) Split(ReadOnlySpan<byte> data)
    {
        var separatorIndex = data.IndexOf(HeaderTerminator);
        if (separatorIndex < 0)
        {
            throw new HttpCodecException(HttpCodecError.Incomplete);
        }

        return (
            data[..separatorIndex].ToArray(),
            data[(separatorIndex + HeaderTerminator.Length)..].ToArray());
    }

    private static string[] ReadHeadLines(byte[] head)
    {
        string text;
        try
        {
            text = new UTF8Encoding(false, true).GetString(head);
        }
        catch (DecoderFallbackException)
        {
            throw new HttpCodecException(HttpCodecError.InvalidFormat);
        }

        var lines = text.Split("\r\n", StringSplitOptions.None);
        if (lines.Length == 0 || lines[0].Length == 0)
        {
            throw new HttpCodecException(HttpCodecError.InvalidFormat);
        }

        return lines;
    }

    private static Dictionary<string, string> ParseHeaders(IEnumerable<string> lines)
    {
        var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var line in lines)
        {
            var separator = line.IndexOf(": ", StringComparison.Ordinal);
            if (separator > 0)
            {
                headers[line[..separator].ToLowerInvariant()] = line[(separator + 2)..];
            }
        }

        return headers;
    }

    private static byte[] ReadBody(ReadOnlySpan<byte> body, IReadOnlyDictionary<string, string> headers)
    {
        if (!headers.TryGetValue("content-length", out var value))
        {
            if (body.Length > MaxBodySize)
            {
                throw new HttpCodecException(HttpCodecError.PayloadTooLarge);
            }

            return body.ToArray();
        }

        if (!int.TryParse(value, out var contentLength) || contentLength < 0)
        {
            throw new HttpCodecException(HttpCodecError.InvalidFormat);
        }

        if (contentLength > MaxBodySize)
        {
            throw new HttpCodecException(HttpCodecError.PayloadTooLarge);
        }

        if (body.Length < contentLength)
        {
            throw new HttpCodecException(HttpCodecError.Incomplete);
        }

        return body[..contentLength].ToArray();
    }

    private static byte[] EncodeHeadAndBody(
        string startLine,
        IReadOnlyDictionary<string, string> headers,
        byte[] body)
    {
        var builder = new StringBuilder(startLine).Append("\r\n");
        foreach (var (key, value) in headers)
        {
            builder.Append(key).Append(": ").Append(value).Append("\r\n");
        }

        var head = Encoding.UTF8.GetBytes(builder.Append("\r\n").ToString());
        var result = new byte[head.Length + body.Length];
        head.CopyTo(result, 0);
        body.CopyTo(result, head.Length);
        return result;
    }

    private static string ReasonPhrase(int statusCode) => statusCode switch
    {
        200 => "OK",
        400 => "Bad Request",
        401 => "Unauthorized",
        403 => "Forbidden",
        404 => "Not Found",
        413 => "Payload Too Large",
        500 => "Internal Server Error",
        _ => "Status",
    };
}
